### PR TITLE
Feature/UI

### PR DESCRIPTION
**Pull request by HenrikThoroe**
This PR adds a simple UI framework. 
Features of the framework are:
- Streamlined way to create CL interfaces
- Reactive state and automatic incremental UI updates
- Support for build in UI elements (`Widget`)
  - `Label` -> styled text
  - `ProgressBar` -> self explaining
  - `Line` -> Combines multiple widgets in one line with optional separators
  - `Table` -> self explaining
- Widgets use one line each (except `ContainerWidget`)
- Toolchain for creating new widgets

### CODE REVIEW

File: lib/core/console/bold.dart [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/// Extension on [String] to convert it to bold text in the terminal.
+extension Bold on String {
+  /// Adds the required ANSI codes to display bold text in the terminal.
+  String bold() {
```
Comment (Coronon) [inline lib/core/console/bold.dart]: Maybe make this a getter? The name 'bold' sounds more like a getter than a method :) But I really like the idea!

File: lib/core/console/cursor.dart [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import 'dart:io';
+
+/// Writes a sequence to [sink] or [stdout] which
+/// moves the cursor to the start of the previous line.
+void moveUp(int lines, {StringSink? sink}) {
```
Comment (Coronon) [inline lib/core/console/cursor.dart:5]: So what does the `lines` argument do then? (Same goes for `moveDown`)

File: lib/core/console/displayed_length.dart [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/// Removes all ANSI sequences from [str] and returns it's resulting length
+int getDisplayedLength(String str) {
+  return str.replaceAll(RegExp('\x1B[[0-9]+[a-zA-Z]'), '').length;
+}
+
+extension DisplayedLength on String {
+  /// The length of this [String] without ANSI escape sequences
+  int get displayedLength => getDisplayedLength(this);
```
Comment (Coronon) [inline lib/core/console/displayed_length.dart:8]: Why implement this as a proxy if its only for String? I would expect this for some templated function.

File: lib/features/ui/render/scene.dart [hunk 4]
```diff
@@ -0,0 +1,97 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'package:flyde/core/console/cursor.dart';
```
Comment (Coronon) [inline lib/features/ui/render/scene.dart]: Relative imports inside _lib_ ;)

File: lib/features/ui/render/scene.dart [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+    for (final widget in _widgets) {
+      widget.line = line++;
+
+      if (widget is StatefulWidget) {
+        widget.updateRequestReceiver = _updateController.sink;
+      }
+
+      if (widget is InlineWidget) {
+        widget.syncBody();
```
Comment (Coronon) [inline lib/features/ui/render/scene.dart:86]: Have you not tested `InlineWidget`s?

File: lib/features/ui/render/scene.dart [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+import 'package:flyde/core/console/cursor.dart';
+import 'package:flyde/features/ui/render/widget.dart';
+
+/// A [Scene] is the default renderer for [Widget]s.
+///
+/// [Scene] renders and displays [Widget]s on the
+/// standard output. For [Scene] to work properly
+/// [stdout] has to have a terminal attached.
+class Scene {
```
Comment (Coronon) [inline lib/features/ui/render/scene.dart:13]: I really like how you implemented this, great work!!!

File: lib/features/ui/render/widget.dart [hunk 7]
```diff
@@ -0,0 +1,142 @@
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+
```
Comment (Coronon) [inline lib/features/ui/render/widget.dart:4]: Wonderful concept! This really reminds me of how Flutter works, we could possibly consider moving this to its own package, just like **license_gatherer** so that we can reuse it in other projects :+1:

File: lib/features/ui/widgets/label.dart [hunk 8]
```diff
@@ -0,0 +1,42 @@
+import 'package:flyde/core/console/terminal_color.dart';
```
Comment (Coronon) [inline lib/features/ui/widgets/label.dart]: Importinos :P

File: lib/features/ui/widgets/line.dart [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import 'dart:math';
+
+import 'package:flyde/core/console/displayed_length.dart';
```
Comment (Coronon) [inline lib/features/ui/widgets/line.dart]: You know the drill ^^

File: lib/features/ui/widgets/line.dart [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  /// The minimum width of each children.
+  final int _width;
+
+  /// The alignment of each child which has a width less than the minimum width
+  final TextAlignment _alignment;
+
+  Line(
+    List<Widget> body,
+    this._separator, {
+    int width = -1,
```
Comment (Coronon) [inline lib/features/ui/widgets/line.dart]: Do we have to use a magic number?

File: lib/features/ui/widgets/progress_bar.dart [hunk 11]
```diff
@@ -0,0 +1,38 @@
+import 'package:flyde/features/ui/render/widget.dart';
```
Comment (Coronon) [inline lib/features/ui/widgets/progress_bar.dart]: \***cough**\* \***cough**\*

File: lib/features/ui/widgets/spacer.dart [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import 'dart:math';
+
+import 'package:flyde/core/console/terminal_color.dart';
```
Comment (Coronon) [inline lib/features/ui/widgets/spacer.dart]: 🤪

File: lib/features/ui/widgets/spacer.dart [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+class Spacer extends Widget with ContainerWidget {
+  /// Number of lines which should be empty.
+  final int _lines;
+
+  Spacer([this._lines = 1]);
+
+  @override
+  List<Widget> get children => List.filled(
+        max(_lines, 1),
+        Label.constant('', color: TerminalColor.none),
```
Comment (Coronon) [inline lib/features/ui/widgets/spacer.dart:17]: I really like how you reuse other widgets and compose them :)

File: lib/features/ui/widgets/table.dart [hunk 14]
```diff
@@ -0,0 +1,105 @@
+import 'package:flyde/core/console/text_alignment.dart';
```
Comment (Coronon) [inline lib/features/ui/widgets/table.dart]: Ok, I think you get the point ^^ I'll not comment on any other imports like this from now on :P

File: lib/features/ui/widgets/table.dart [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+///      Label.constant('Child2.3'),
+///    ]
+///  ],
+///);
+/// ```
+///
+/// renders to
+///
+///```bash
+///                  Item1                  │                   Item1                  │                   Item1
```
Comment (Coronon) [inline lib/features/ui/widgets/table.dart]: Looks a little scuffed on GitHub (Half screen), but I like that you included a practical example (keep it!) :+1:

File: test/features/ui/render/scene_test.dart [hunk 16]
```diff
@@ -0,0 +1,86 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'package:flyde/features/ui/render/scene.dart';
+import 'package:flyde/features/ui/render/widget.dart';
+import 'package:test/test.dart';
+
+import '../../../helpers/mocks/mock_widget.dart';
```
Comment (Coronon) [inline test/features/ui/render/scene_test.dart:10]: But here you import correctly lmao :P

I know that I promised "no more import comments" but I had to, I couldn't fight the urge xD

File: test/features/ui/render/scene_test.dart [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+      );
+    });
+
+    scene.show();
+    await hook.awaitValue(
+      timeout: Duration(seconds: 1),
+      condition: (cnt) => cnt.length == 4,
+    );
+
+    hook.expect(hasLength(4));
```
Comment (Coronon) [inline test/features/ui/render/scene_test.dart]: Why? If `hook.awaitValue` did not throw you can be sure that the length is `4`

File: test/features/ui/render/widget_test.dart [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+
+  test('Renders expected output', () {
+    expect(widget.render(), equals('mock-$_initialPrimaryState'));
+  });
+
+  test('Requests update on state change', () async {
+    final hook = VHook.empty();
+
+    controller.stream.listen((event) {
+      expect(event.line, equals(0));
```
Comment (Coronon) [inline test/features/ui/render/widget_test.dart]: You could complete with a value and then wait for the value with `awaitValue` rather then  `awaitCompletion` :)

Your call

File: test/features/ui/widgets/progress_bar_test.dart [hunk 19]
```diff
@@ -0,0 +1,83 @@
+import 'dart:async';
+
+import 'package:flyde/features/ui/render/widget.dart';
+import 'package:flyde/features/ui/widgets/progress_bar.dart';
+import 'package:test/test.dart';
```
Comment (Coronon) [inline test/features/ui/widgets/progress_bar_test.dart]: import order and seperation

File: test/features/ui/widgets/spacer_test.dart [hunk 20]
```diff
@@ -0,0 +1,24 @@
+import 'package:flyde/features/ui/widgets/label.dart';
+import 'package:flyde/features/ui/widgets/spacer.dart';
+import 'package:test/test.dart';
```
Comment (Coronon) [inline test/features/ui/widgets/spacer_test.dart]: order

File: test/features/ui/widgets/table_test.dart [hunk 21]
```diff
@@ -0,0 +1,257 @@
+import 'package:flyde/core/console/text_alignment.dart';
+import 'package:flyde/features/ui/render/widget.dart';
+import 'package:flyde/features/ui/widgets/table.dart';
+import 'package:test/test.dart';
```
Comment (Coronon) [inline test/features/ui/widgets/table_test.dart]: Can we stick with the order:
```
External

lib (with package: because we are in the tests dir)

test (relative)
```

File: test/features/ui/widgets/table_test.dart [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+    );
+    expect(
+      displayedChildren[4].render(),
+      equals(
+        '             body2.1$verticalSpacer             body2.2$verticalSpacer             body2.3',
+      ),
+    );
+  });
+
+  test('Creates expected children with different cell width', () {
```
Comment (Coronon) [inline test/features/ui/widgets/table_test.dart:177]: What happens if `cellWidth < ContentSize`?

**Review by Coronon — CHANGES_REQUESTED** [review]
I really have to compliment you for this one! I love the clean, well-structured and composed solutions you came up with and, as I've also noted in one of my comments, would consider moving this into its own package.

A few minor tweaks here and there and this should be ready to land :)

File: lib/core/console/cursor.dart (hunk 2 above)
Comment (HenrikThoroe) [reply lib/core/console/cursor.dart:5]: I updated the docs, please have a look and resolve if you think they are ok now

File: lib/core/console/displayed_length.dart (hunk 3 above)
Comment (HenrikThoroe) [reply lib/core/console/displayed_length.dart:8]: It's a proxy because I need it as a function at some point (Label.dart) to pass it to `map`.
On any other occasion an extension method is much cleaner imo. It's used in both ways.
That's for the why, if you think we should only have on way to access the function, let me know

File: lib/features/ui/render/scene.dart (hunk 5 above)
Comment (HenrikThoroe) [reply lib/features/ui/render/scene.dart:86]: I did, have a look at `widget_test.dart`. Or did I miss your question somehow?

File: lib/features/ui/render/scene.dart (hunk 6 above)
Comment (HenrikThoroe) [reply lib/features/ui/render/scene.dart:13]: thanks :)

File: lib/features/ui/render/widget.dart (hunk 7 above)
Comment (HenrikThoroe) [reply lib/features/ui/render/widget.dart:4]: It's an possibility for sure. I'd prefer to "battle test" it in feature/cli first and then possibly move it to its own package. What are your thoughts about that?

File: lib/features/ui/widgets/spacer.dart (hunk 13 above)
Comment (HenrikThoroe) [reply lib/features/ui/widgets/spacer.dart:17]: Thanks :)

File: lib/features/ui/widgets/table.dart (hunk 15 above)
Comment (HenrikThoroe) [reply lib/features/ui/widgets/table.dart]: Updated to 20 cols for each item.

File: test/features/ui/render/scene_test.dart (hunk 17 above)
Comment (HenrikThoroe) [reply test/features/ui/render/scene_test.dart]: For better readability of the test. I think it's easier to see that we expect four item when it's an explicit test case. Resolve if you agree otherwise write 'No more four' so I remove it :)

File: test/features/ui/widgets/table_test.dart (hunk 22 above)
Comment (HenrikThoroe) [reply test/features/ui/widgets/table_test.dart:177]: It looks insanely ugly xD 

What happens is, that `Line`, which is the foundation for each row containing cells, uses the children width and does not apply any padding. The horizontal separators on the other side scale with `cellWidth` and are shorter than the content. 
A fix could be to pre-render the table content, which would violate the paradigm of `ContainerWidget`.
Because items could and will have different widths a check for the required width of the separators must include a complete render of all children. I'm not sure if that's worth it. 

TL;DR You shouldn't do that (trademark) (copyright) (what could possibly go wrong?)

PS: Thanks for the point out. Didn't think of this issue before. Maybe open a GitHub issue for that and let it rod?

File: test/features/ui/widgets/table_test.dart (hunk 21 above)
Comment (HenrikThoroe) [reply test/features/ui/widgets/table_test.dart]:
```
dart 

external (e.g. test)

lib

relative imports
```

I don't want to mix `dart:` imports with `package:` imports

Comment (HenrikThoroe) [reply test/features/ui/widgets/table_test.dart]: I resolve the other related comments to keep the review a little cleaner

Comment (Coronon) [reply test/features/ui/widgets/table_test.dart]: Jup, simply forgot about the dart imports, thats how I do it already 👍🏻

File: lib/features/ui/render/scene.dart (hunk 5 above)
Comment (Coronon) [reply lib/features/ui/render/scene.dart:86]: I saw that, but it seems to not interact with a `Scene` in a way that would run `_setup()`. I am trying to work out why codecov complains about line 85 not being tested.

File: lib/core/console/displayed_length.dart (hunk 3 above)
Comment (Coronon) [reply lib/core/console/displayed_length.dart:8]: Makes perfect sense then :)

File: lib/features/ui/render/widget.dart (hunk 7 above)
Comment (Coronon) [reply lib/features/ui/render/widget.dart:4]: Sounds like a plan. Lets see if it works as expected first before going to the lengths of seperating it from the main code base. On the other hand the friction in import migration would be less (but rewriting a few imports should be fine.

File: test/features/ui/render/scene_test.dart (hunk 17 above)
Comment (Coronon) [reply test/features/ui/render/scene_test.dart]: You are right, readability matters with test cases. Furthermore we dont have to consider performance at all here, so yes, keep it :)

File: test/features/ui/widgets/table_test.dart (hunk 22 above)
Comment (Coronon) [reply test/features/ui/widgets/table_test.dart:177]: Wellll.... ^^

Can't we check the width after already rendering like the following?:
1. Render everything
2. Check that constraints are not violated (raise exception on violation)

File: lib/features/ui/render/scene.dart (hunk 5 above)
Comment (HenrikThoroe) [reply lib/features/ui/render/scene.dart:86]: The issue is that the scene test has no inline widget in the widget list. The `syncBody` method is fully tested though.

File: test/features/ui/widgets/table_test.dart (hunk 22 above)
Comment (HenrikThoroe) [reply test/features/ui/widgets/table_test.dart:177]: Finally I got some time for implementing the feature. You can find the tests in `scene_test.dart`

Comment (HenrikThoroe) [reply test/features/ui/widgets/table_test.dart:177]: For table: I added a test which ensures that each child has min and max width set. The enforcing of those constraints is a job for `Scene` and therefore not tested in `table_test`

File: lib/features/ui/render/scene.dart (hunk 5 above)
Comment (HenrikThoroe) [reply lib/features/ui/render/scene.dart:86]: Do you want me to include a `InlineWidget` in the list of widgets for scene test? Please resolve otherwise

Comment (Coronon) [reply lib/features/ui/render/scene.dart:86]: I would love to achive max. coverage, so if its not to much work, yes please :)

File: test/features/ui/widgets/table_test.dart (hunk 22 above)
Comment (Coronon) [reply test/features/ui/widgets/table_test.dart:177]: Well done!

**Comment by Coronon** [comment]
Basically ready to merge, but would you mind making this perfect with fixing these 4 lines not being covered?:

https://coronon.github.io/flyde/feature/ui/features/ui/render/scene.dart.gcov.html

https://coronon.github.io/flyde/feature/ui/features/ui/widgets/label.dart.gcov.html

**Review by Coronon — APPROVED** [review]
